pkg/operation/botanist/dnsresources_test.go

```diff
@@ -0,0 +1,301 @@
+									Addon: gardencorev1beta1.Addon{
+										Enabled: true,
+									},
+								},
+							},
+						},
+					},
+					ExternalClusterDomain: pointer.String(externalDomain),
+					ExternalDomain: &garden.Domain{
+						Provider: externalProvider,
+					},
+					Components: &shoot.Components{
+						Extensions: &shoot.Extensions{
+							DNS: &shoot.DNS{
+								ExternalOwner:    externalDNSOwner,
+								ExternalProvider: externalDNSProvider,
+								ExternalEntry:    externalDNSEntry,
+								InternalOwner:    internalDNSOwner,
+								InternalProvider: internalDNSProvider,
+								InternalEntry:    internalDNSEntry,
+								NginxOwner:       ingressDNSOwner,
+								NginxEntry:       ingressDNSEntry,
+							},
+							ExternalDNSRecord: externalDNSRecord,
+							InternalDNSRecord: internalDNSRecord,
+							IngressDNSRecord:  ingressDNSRecord,
+						},
+					},
+				},
+				Garden: &garden.Garden{
+					InternalDomain: &garden.Domain{
+						Provider: internalProvider,
+					},
+				},
+				Logger: logrus.NewEntry(logger.NewNopLogger()),
+			},
+		}
+	})
+
+	AfterEach(func() {
+		ctrl.Finish()
+	})
+
+	Describe("#DeployInternalDNSResources", func() {
+		It("should delete the DNSOwner, DNSProvider, and DNSEntry resources, and then deploy the DNSRecord resource if the feature gate is enabled", func() {
+			defer test.WithFeatureGate(gardenletfeatures.FeatureGate, features.UseDNSRecords, true)()
+			gomock.InOrder(
+				internalDNSOwner.EXPECT().Destroy(ctx),
+				internalDNSProvider.EXPECT().Destroy(ctx),
+				internalDNSEntry.EXPECT().Destroy(ctx),
+				internalDNSRecord.EXPECT().Deploy(ctx),
+				internalDNSRecord.EXPECT().Wait(ctx),
+			)
+			Expect(b.DeployInternalDNSResources(ctx)).To(Succeed())
+		})
+
+		It("should migrate and delete the DNSRecord resource, and then deploy the DNSOwner, DNSProvider, and DNSEntry resources if the feature gate is disabled", func() {
+			defer test.WithFeatureGate(gardenletfeatures.FeatureGate, features.UseDNSRecords, false)()
+			gomock.InOrder(
+				internalDNSRecord.EXPECT().Migrate(ctx),
+				internalDNSRecord.EXPECT().WaitMigrate(ctx),
+				internalDNSRecord.EXPECT().Destroy(ctx),
+				internalDNSRecord.EXPECT().WaitCleanup(ctx),
+				internalDNSOwner.EXPECT().Deploy(ctx),
+				internalDNSOwner.EXPECT().Wait(ctx),
+				internalDNSProvider.EXPECT().Deploy(ctx),
+				internalDNSProvider.EXPECT().Wait(ctx),
+				internalDNSEntry.EXPECT().Deploy(ctx),
+				internalDNSEntry.EXPECT().Wait(ctx),
+			)
+			Expect(b.DeployInternalDNSResources(ctx)).To(Succeed())
+		})
+	})
+
+	Describe("#DeployExternalDNSResources", func() {
+		It("should delete the DNSOwner, DNSProvider, and DNSEntry resources, and then deploy the DNSRecord resource if the feature gate is enabled", func() {
+			defer test.WithFeatureGate(gardenletfeatures.FeatureGate, features.UseDNSRecords, true)()
+			gomock.InOrder(
+				externalDNSOwner.EXPECT().Destroy(ctx),
+				externalDNSProvider.EXPECT().Destroy(ctx),
```

Comment:
The external `DNSProvider` must not be deleted as it may still be needed for `DNSEntries` of the `shoot-dns-service`. The internal `DNSProvider` can be deleted, but not the external one.

Reply:
If I understand this correctly, if it can be used for `DNSEntries` of the `shoot-dns-service`, not only should it not be deleted, but it should be also **created** if a new cluster is created with `UseDNSRecords=true`, correct? So far I thought that such a `DNSProvider` should be created for the "primary" provider listed in `spec.dns.providers`, if there is such a provider, but not if there is no such provider and a "default domain" secret is used. Should it really also be created for the "default domain" secret case as well?

@timebertt @rfranzke Could you please help here? I am not sure I fully understand what exactly is supported right now. Let's clarify this before I do further changes.

Reply:
My thoughts on the matter:

* If the above assumption is true, then the external `DNSProvider` should always be created (and not touched when migrating from `UseDNSRecords=true` to `UseDNSRecords=false` and vice versa), either from the "primary" provider in the shoot spec or from the "default domain" secret. This however is contradictory to the idea to completely decouple Gardener from `DNSProviders` in the future, so we should think carefully if we really need this.
* If the above assumption is not true, then the  external `DNSProvider` must only be created if there is a "primary" provider in the shoot spec. In the migration case, the above logic may delete it, but it will still be created as an "additional provider" later and will therefore still exist after the migration. I assume this should be ok in this case?

Reply:
I think, you understood correctly, @stoyanr.
However, @MartinWeindel has a good point here (nice catch, didn't think about this case in my reviews): currently, the default dns provider can be used by the shoot-dns-service for DNSEntries in the shoot as well, because of the implicit dependencies involved here.
So we probably indeed need to keep/reconcile the `DNSProvider` for the default domain as well for backwards-compatibility.

Reply:
Hm, I couldn't follow this in detail from the comments, but why do we have to keep this logic in gardenlet? Can't we rather change the `shoot-dns-service` extension to take over responsibility for the "external" `DNSProvider`? In the migration case (`UseDNSRecords=true`) we could simply not delete it?

Reply:
This is my understanding about shifting the responsibility for the `DNSProvider` to the `shoot-dns-service` after a private chat with @stoyanr:
- it should be dealt with only after the dust of the migration to `DNSRecord` has settled 
- the specification of the DNS providers should be migrated from the `spec.dns.providers` section in the shoot manifest to the `spec.extensions[type=shoot-dns-service].providerConfig.providers` section
- the provider secrets located in the virtual garden must be made available to the extension, e.g. by copying them to the CP using "referenced resources"

Reply:
> This is my understanding about shifting the responsibility for the `DNSProvider` to the `shoot-dns-service` after a private chat with @stoyanr:

Yes, I agree, this is also what I had in mind. We will need a second migration step for the `DNSProvider`s in order to resolve this implicit coupling between gardener and shoot-dns-service.
But let's focus on the `DNSRecord`s only for now, not touch the `DNSProvider`s and take care about the shift for the `DNSProvider`s later on 👍

Reply:
As agreed, I added logic to ensure that the external DNSProvider is always created (also when `UseDNSRecords=true`), and removed the special handling of the primary DNS provider in `AdditionalDNSProviders` that was previously introduced. Still testing all possible cases.

Reply:
I tested the following cases and it seems to work ok:
* No DNS providers in shoot spec - the external `DNSProvider` is still created from the default-domain secret
* One non-primary DNS provider - both the external `DNSProvider` and the non-primary `DNSProvider` are created.
* One primary DNS provider - only the primary `DNSProvider` is created from the info in the shoot spec.
I also tested hibernation, CP migration, etc. for the third case.